src/transformStateWithClones.js

```diff
@@ -7,7 +7,48 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let referenceState = { ...state };
+
+  for (const action of actions) {
+    const copy = { ...referenceState };
+    const { type } = action;
+
+    switch (type) {
+      case 'addProperties': {
+        const { extraData } = action;
+
+        for (const [key, value] of Object.entries(extraData)) {
+          copy[key] = value;
+        }
+
+        stateHistory.push(copy);
+        referenceState = { ...copy };
```

Comment:
[DONT REPEAT YOURSELF] - The action of pushing the 'copy' to 'stateHistory' and updating 'referenceState' is repeated in each case. Consider performing this action once after the switch statement to adhere to the DRY principle.